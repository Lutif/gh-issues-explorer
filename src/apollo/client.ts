import {
  ApolloClient,
  ApolloLink,
  HttpLink,
  InMemoryCache,
} from "@apollo/client";
import { onError } from "@apollo/client/link/error";
import { REACT_APP_GITHUB_BASE_URL } from "../utils/constants";

const httpLink = new HttpLink({
  uri: REACT_APP_GITHUB_BASE_URL,
  headers: {
    authorization: `Bearer ${process.env.REACT_APP_GITHUB_PERSONAL_ACCESS_TOKEN}`,
  },
});

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path }) =>
      //can add loging here, eg: sentry
      console.error(
        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );
  }
  if (networkError) {
    const hasErrorCode = networkError.name === "Unauthorised";
    if (hasErrorCode) {
      //do something
    }
    console.log(`[Network error]: ${networkError}`);
  }
});

const link = ApolloLink.from([errorLink, httpLink]);

const cache = new InMemoryCache({
  typePolicies: {
    Issue: {
      keyFields: ["id"],
    },
    User: {
      keyFields: ["login"],
    },
  },
});

export const client = new ApolloClient({
  link,
  cache,
});
