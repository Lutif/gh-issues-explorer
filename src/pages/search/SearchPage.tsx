import { Button } from "@mui/material";
import React, { useState } from "react";
import { IssueListItem } from "./components/IssueListItem";
import { SearchBar } from "./components/Searchbar";
import { PAGE_SIZE, useSearchIssues } from "../../utils";
import { ShouldRender, Loading } from "../shared";
import Checkbox from "@mui/material/Checkbox";
import { SearchPageStyled } from "./searchPage.styles";
import FormGroup from "@mui/material/FormGroup";
import FormControlLabel from "@mui/material/FormControlLabel";

interface SearchPageProps {}

export const SearchPage: React.FC<SearchPageProps> = () => {
  const [open, setOpen] = useState(true);
  const {
    search,
    gotoNextPage,
    gotoPrevPage,
    data: issues,
    currentPage,
    hasMore,
    isLoading,
    setSearchText,
  } = useSearchIssues();

  const performSearch = () => {
    search({ isOpen: open });
  };

  return (
    <SearchPageStyled>
      <div className="header">
        <SearchBar
          setSearchText={setSearchText}
          performSearch={performSearch}
        />
        <FormGroup className="status-checkboxes">
          <FormControlLabel
            control={
              <Checkbox
                checked={open}
                onClick={() => setOpen((prev) => !prev)}
              />
            }
            label="Open"
          />
        </FormGroup>
        <Button
          onClick={performSearch}
          className="search-btn"
          variant="contained"
          color="primary"
        >
          Search
        </Button>
      </div>
      <ShouldRender if={isLoading}>
        <Loading />
      </ShouldRender>
      <ShouldRender if={!isLoading}>
        <>
          <ShouldRender if={issues.length > 0}>
            <div className="issues">
              {issues?.map((issue, index) => (
                <IssueListItem key={`${issue.id}-${index}`} issue={issue} />
              ))}
            </div>
          </ShouldRender>
          <ShouldRender if={issues.length === 0}>
            <div className="no-result">No result found</div>
          </ShouldRender>
        </>
      </ShouldRender>
      <ShouldRender if={issues.length > 0}>
        <div className="pagination-container">
          <Button
            variant="outlined"
            color="info"
            size="small"
            disabled={currentPage <= 1 || isLoading}
            onClick={() => gotoPrevPage()}
          >
            prvious page
          </Button>
          <p className="page">{currentPage}</p>
          <Button
            variant="outlined"
            color="info"
            size="small"
            disabled={(!hasMore && issues.length < PAGE_SIZE) || isLoading}
            onClick={() => gotoNextPage(open)}
          >
            Next page
          </Button>
        </div>
      </ShouldRender>
    </SearchPageStyled>
  );
};
