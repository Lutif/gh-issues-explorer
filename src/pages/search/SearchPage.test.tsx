import { render, screen } from "@testing-library/react";
import { SearchPage } from "./SearchPage";
import { MemoryRouter } from "react-router-dom";
import { MockedProvider } from "@apollo/client/testing";

describe("<SearchPage /> tests", () => {
  it("renders the SearchPage component", () => {
    render(
      <MemoryRouter>
        <MockedProvider>
          <SearchPage />
        </MockedProvider>
      </MemoryRouter>
    );
    const element = screen.getByText("Search");
    expect(element).toBeInTheDocument();
  });

  it("test SearchPage snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MockedProvider>
          <SearchPage />
        </MockedProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
