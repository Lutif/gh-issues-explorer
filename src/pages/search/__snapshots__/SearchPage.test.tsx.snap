// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPage /> tests test SearchPage snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dkzDqf dCLrAF"
  >
    <div
      class="header"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        style="width: 100%;"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            placeholder="Search issue title or body"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormGroup-root status-checkboxes css-dmmspl-MuiFormGroup-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Open
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Close
          </span>
        </label>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium search-btn css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Search
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="sc-gsnTZi cApuSV"
      data-testid="loading-spinner-test-id"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
      ;
    </div>
    <div
      class="pagination-container"
    >
      <button
        disabled=""
      >
        prvious page
      </button>
      <p
        class="page"
      >
        0
      </p>
      <button
        disabled=""
      >
        Next page
      </button>
    </div>
  </div>
</DocumentFragment>
`;
